tclpro/modules/debugger/doc/prodebug.1
'\"
'\" Copyright (c) 1998-2000 by Ajuba Solutions
'\" All rights reserved.
'\"
'\" RCS: @(#) $Id: prodebug.1,v 1.2 2000/08/08 02:06:33 welch Exp $
'\" 
.so man.macros
.TH prodebug 1 "" TclPro "TclPro Applications"
.BS
'\" Note:  do not modify the .SH NAME line immediately below!
.SH NAME
prodebug \- TclPro Debugger

.SH SYNOPSIS
\fBprodebug\fR ?\fB\-help\fR? ?\fB\-nologo\fR? ?\fB\-coverage\fR? ?\fB\-display\fR \fIdisplay\fR? ?projectFile?

.SH OPTIONS
.PP
.IP "\fB\-help\fR" 15
Displays usage information, then exit without doing anything.
.IP "\fB\-nologo\fR" 15
Hides the copyright and version information. 
.IP "\fB\-coverage\fR" 15
Enables the code coverage analysis feature.
.IP "\fB\-display\fR \fIdisplay\fR" 15
Display (and screen) on which to display window.  (UNIX only.)
.BE

.SH DESCRIPTION
.PP

TclPro Debugger is a graphical debugger for the Tcl scripting
language.  TclPro Debugger allows you to interactively step through
your Tcl scripts to find bugs as they occur.  You may also use the
TclPro Debugger to modify the state of your running Tcl application to
test hard-to-execute or otherwise unreachable code branches.

.PP

The command \fBprodebug\fR is used to start TclPro Debugger.  If you
specify a \fIprojectFile\fR on the command line, TclPro Debugger will
open the project upon startup.

.SH "USER MANUAL"
.PP
For more information, please see the TclPro User's Guide or online
help system.

.PP
The code coverage feature is a late addition to the tool and is
not documented in the user manual.  If you start \fBprodebug\fR
with the \fB-coverage\fR option then a "Show Coverage" menu entry appears under
the View menu, and a toolbar button labeled \fBC\fR appears.  These
display a simple code coverage tool that will highlight either
the code that has been executed so far, or the code that has
not been executed so far.  The idea of this tool is to show you
how well test inputs are exercising your program.


.SH KEYWORDS
debug


tclpro/ChangeLog
2001-03-31  Andreas Kupries <andreas_kupries@users.sourceforge.net>

	* get-trfcrypt: Shortened in adaption to the new release of
	  Trfcrypt (2.0p3).

2001-03-31  Andreas Kupries <andreas_kupries@users.sourceforge.net>

	* get-setops: Shortened in adaption to the new release of setops
	  (1.0p2).

2001-03-28  Andreas Kupries <andreas_kupries@users.sourceforge.net>

	* get-trf:
	* get-memchan: Shortened in adaption to the new releases of trf
	  and memchan.

2001-03-16  Andreas Kupries <andreas_kupries@users.sourceforge.net>

	* get-trf:
	* get-expand:
	* get-trfcrypt:
	* get-setops:
	* get-memchan: Added code to move an existing archive before retrieving
	  the current one from the web. Without this 'wget' will create
	  names like 'foo.1', etc. instead of 'foo'. This causes the code
	  after the wget to actually use the old archive and not the new one.

2001-03-15  Karl Lehenbauer <karl@procplace.com>
	* snapshot/{README bundle.tcl listpackages.tcl postpro.tcl}:
	  Added snapshot directory and these new files.  

	  postpro.tcl grinds a SNAPSHOT file to produce a SNAPSHOT.merge 
	  file, containing concise information about what modules produced
	  what files in the output and defined what packages.

	  bundle.tcl takes a SNAPSHOT.merge file and produces numerous
	  zip files consisting of the files for each module, with
	  manual pages and demos split into separate zips.

	  listpackage.tcl if run by a tcl or wish shell will list all
	  of the packages available to that shell.

2001-03-14  Karl Lehenbauer <karl@procplace.com>

	* mkIndex.tcl.in: Altered Nativepath proc to work correctly with 
	  new Cygwin drive specification syntax when running on Windows.
	  (Was //d/foo -> D:/foo   Now also /cygdrive/d/foo -> D:/foo)

2001-03-07 Karl Lehenbauer <karl@procplace.com>
	* modules/install/src/setup.sh: Removed expectation that
	  uname resides in /bin -- it doesn't on FreeBSD and probably
	  others.

2001-03-06 Andreas Kupries <andreas_kupries@users.sourceforge.net>

	* configure.in:
	* module_data.tcl.in
	* modules/install/src/parts.in:
	* modules/install/src/cdparts.in: Added information to allow the
	  distribution of the tcl-based macro processor 'expand' by
	  William H. Duquette <will@wjduquette.com>. See
	  http://www.wjduquette.com/expand/index.html

	* get-expand: New file. Retrieves expand2.0 and patches the
	  distribution for usage by the huild framework.

2001-03-06 Andreas Kupries <andreas_kupries@users.sourceforge.net>

	* configure.in:
	* module_data.tcl.in
	* modules/install/src/parts.in:
	* modules/install/src/cdparts.in: Added information to allow the
	  compilation of 'memchan', 'trf' and 'trfcrypt'.

	* get-trf:
	* get-trfcrypt:
	* get-memchan: New files. Retrieve the current source
	  distributions of memchan, trf and trfcrypt, and patch them so
	  that the build framework can use them.

	* README.akupries: New file, short explanation of the above.

2001-02-23 Karl Lehenbauer  <karl@procplace.com>
	Documented and simplified soft status code in buildenv/buildModule.tcl.
	It is only needed for caching that configure has been done and
	whether or not "make install" has been done.  Now that's all
	we try to track.

	Added new make target, "make reinstall".  If you do this, it
	will build and install everything, ignoring any soft status
	information.

2001-02-20 Karl Lehenbauer  <karl@procplace.com>
	Modified dependency definitions in tclpro/module_data.tcl.in 
	and tclchecker/DEPENDENCIES.  In particular, tclchecker needs
	prowrap.  Noted in comments in module_data.tcl.in which modules'
	dependencies are overridden by explicit DEPENDENCIES files.

	Modified buildenv/buildModule.tcl and tclpro/Makefile.in to
	add the capability of skipping steps that have already been
	successfully performed by caching results in a SOFTSTATUS file
	managed by buildModule.tcl in the directory you run "make"
	from to build tclpro.  Thus if, for example, TclX has already
	been installed, it isn't installed over and over again whenever
	you do a make.  This is a little kludgey because it works by
	having, for example, the successful return of "make install" of
	TclX letting us know it worked, rather than looking for a specific
	file result.

	Also buildModule exits upon hitting an error, rather than proceeding.
	This should probably be made an option.

	Finally all TEA supported targets are now supported from the
	Makefile, so you can say stuff like "make tcl-install", 
	"make tcl_static-install", "make tcllib-clean", 
	"make tclhttpd-configure" etc.  Using make in this manner bypasses
	the soft status stuff described above.

2001-02-8  Brent Welch  <brent.welch@interwoven.com>
	TclProCheckout.tcl: Cleaned this up so it correctly detects if
	a read/write checkout is successful.  Added -no_action cmdline arg.

2001-02-6  Brent Welch  <brent.welch@interwoven.com>
	modules/install/src/genImage.tcl: Rooted out more references to license
	modules/install/src/text.tcl: Rooted out more references to license

2001-01-30  Brent Welch  <brent.welch@interwoven.com>
	README.FreeBSD: latest updates by Karl
	configure.in: Removed builds of documentation and mclistbox

2001-01-24  Brent Welch  <brent.welch@interwoven.com>
	(For Karl Lehenbauer)
	modules/wrapper/src/setup.tcl:
	Removed references to license code
	README.FreeBSD: added docs from Karl.

2001-01-24  Brent Welch  <brent.welch@interwoven.com>
	(For Karl Lehenbauer)
	modules/wrapper/src/startup.tcl:
	modules/debugger/src/startup.tcl: changed the way the
	    lcient calls are commented out so it shows up from a grep better
	modules/install/src/cdparts.in:
	modules/install/src/genImage.tcl:
	modules/install/src/gui.tcl:
	modules/install/src/install.tcl:
	modules/install/src/makeDist.tcl:
	    Fixes by Karl to get the installer images working on FreeBSD!
	modules/lmgr:
	    Removed the old, unused license manager.

2001-01-22  Brent Welch  <brent.welch@interwoven.com>
	* Makefile.in:
	Added a %-clean target that is equivalent to %-hose -

tclpro/modules/demos/tutorials/prodebug.tcl
# prodebug.tcl --
#
#	This file contains the public routines used to start debugging user
#	code in a remote application.
#
# Copyright (c) 1998-1999 by Scriptics Corporation.
# See the file "license.terms" for information on usage and redistribution of this file.
#
# RCS: @(#) $Id: prodebug.tcl,v 1.2 2000/10/31 23:31:12 welch Exp $


#
# This file comprises the public interface to the TclPro Debugger for use
# by applications that are not launched directly from the debugger.  The
# public interface consists of the two commands "debugger_init" and
# "debugger_eval".  A typical application will source this file then invoke
# "debugger_init" to initiate the connection to the debugger.  Once
# connected, the application can use the "debugger_eval" command to
# evaluate scripts that the debugger will be able to step through.
# Additionally, various other Tcl commands including "source" and "proc"
# will automatically instrument code.  Any blocks of code (e.g. procedure
# bodies) that existed before "debugger_init" was invoked will execute
# without any instrumentation.
#


# debugger_init --
#
#	This function initiates a connection to the TclPro Debugger.  Files
#	that are sourced and procedures that are defined after this
#	function completes will be instrumented by the debugger.
#
# Arguments:
#	host	Name of the host running the debugger.
#	port	TCP port that the debugger is using.
#
# Results:
#	Returns 1 on success and 0 on failure.


proc debugger_init {{host 127.0.0.1} {port 2576}} {
    global tcl_version

    if {[catch {set socket [socket $host $port]}] != 0} {
	return 0
    }
    fconfigure $socket -blocking 1 -translation binary

    # On 8.1 and later versions we should ensure the socket is not doing
    # any encoding translations.

    if {$tcl_version >= 8.1} {
	fconfigure $socket -encoding utf-8
    }

    # Send the loader and tcl library version

    set msg [list HELLO 1.0 $tcl_version]
    puts $socket [string length $msg]
    puts -nonewline $socket $msg
    flush $socket

    # Get the rest of the nub library and evaluate it in the current scope.
    # Note that the nub code assumes there will be a "socket" variable that
    # contains the debugger socket channel.

    if {[gets $socket bytes] == -1} {
	close $socket
	return 0
    }
    set msg [read $socket $bytes]
    eval [lindex $msg 1]
    return 1
}

# debugger_eval --
#
#	Instrument and evaluate a script.  This routine is a trivial
#	implementation that is replaced when the nub is downloaded.
#
# Arguments:
#	args		One or more arguments, the last of which must
#			be the script to evaluate.
#
# Results:
#	Returns the result of evaluating the script.

proc debugger_eval {args} {
    global errorInfo errorCode
    set length [llength $args]
    if {$length < 1} {
	error "wrong # args: should be \"debugger_eval ?options? script\""
    }
    set code [catch {uplevel 1 [lindex $args [expr {$length - 1}]]} result]
    return -code $code -errorcode $errorCode -errorinfo $errorInfo $result
}

# debugger_break --
#
#	This command may be inserted in user code to cause a break 
#	to occur at the location of this command.  If the application
#	is not connected to the debugger this command is a no-op.
#
# Arguments:
#	str	(Optional) String that displays in debugger.
#
# Results:
#	None.  Will send break message to debugger.

proc debugger_break {{str ""}} {
    return
}
